File: S2-Regression/Pt2-MultipleLinearRegression/MultipleRegression.R
### Multipole Linear Regression

## Set current working directory
getwd()
setwd('C:\\Users\\rzl\\Documents\\Github\\Machine-Learning\\S2-Regression\\Pt2-MultipleLinearRegression/')
getwd()

## Import data
dataset = read.csv('Startups.csv')

## Encoding categorical data
dataset$State = factor(dataset$State, levels=c('New York', 'California', 'Florida'), labels=c(1, 2, 3))

## Splitting the dataset into Training set and Test set
# Import caTools
install.packages('caTools')
library('caTools')

# set.seed() -> Set the seed of R's random number generator, which is useful for creating simulations
set.seed(123)

# sample.split() -> Split Data Into Test And Train Set
dataset_split = sample.split(dataset$Profit, SplitRatio=3/4)
training_set = subset(dataset, dataset_split == TRUE)
test_set = subset(dataset, dataset_split == FALSE)

# ## Feature Scaling
# training_set[, 2:3] = scale(training_set[, 2:3])
# test_set[, 2:3] = scale(test_set[, 2:3])

## Fitting Simple Linear Regression to the Training set
regressor = lm(formula=Profit~., data=training_set)
summary(regressor)

## Prediction the Test set results
y_predict = predict(regressor, newdata=test_set)
y_predict

## Steps of Backward Eliminations
# 1. Select a signficance level to stay in the model (e.g. SL = 0.05)
# 2. Fit the full model with all possible predictors
# 3. Consider the predictor with the highest P value. if P > SL, go to step 4, otherwise it's done
# 4. Remove the predictor
# 5. Fit model without his variable

## Building the optimal model using Backward Elimination
# Step 3 - 5
regressor = lm(formula=Profit~R.D.Spend+Administration+Marketing.Spend+State, data=dataset)
summary(regressor)

regressor = lm(formula=Profit~R.D.Spend+Administration+Marketing.Spend, data=dataset)
summary(regressor)

regressor = lm(formula=Profit~R.D.Spend+Marketing.Spend, data=dataset)
summary(regressor)

regressor = lm(formula=Profit~R.D.Spend, data=dataset)
summary(regressor)

## Automatic Backward Elimination
backwardElimination <- function(x, sl) {
  numVars = length(x)
  for (i in c(1:numVars)){
    regressor = lm(formula = Profit ~ ., data = x)
    maxVar = max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
    if (maxVar > sl){
      j = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
      x = x[, -j]
    }
    numVars = numVars - 1
  }
  return(summary(regressor))
}

SL = 0.05
dataset = dataset[, c(1,2,3,4,5)]
backwardElimination(training_set, SL)

File: S2-Regression/Pt2-MultipleLinearRegression/multiple_regression.py
### Data Preprocessing

## Import all the library that needed
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Import dataset with pandas
dataset = pd.read_csv('Startups.csv')

# pandas.DataFrame.iloc >> Purely integer-location based indexing for selection by position.
x = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

## Enoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import ColumnTransformer

## ColumnTransformer >> Applies transformers to columns of an array or pandas DataFrame
columnTransformer = ColumnTransformer([('encoder', OneHotEncoder(), [3])], remainder='passthrough')
x = np.array(columnTransformer.fit_transform(x), dtype=np.float)

## Avoiding the dummy variable trap
x = x[:, 1:]

## Splitting the dataset into Training set and Test set
from sklearn.model_selection import train_test_split

# train_test_split >> Split arrays or matrices into random train and test subsets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=1/4, random_state=0)

# ## Feature Scaling
# from sklearn.preprocessing import StandardScaler

# # StandardScaler() >> Standardize features by removing the mean and scaling to unit variance
# sc = StandardScaler()
# x_train = sc.fit_transform(x_train)
# x_test = sc.transform(x_test)

## Fitting Multiple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(x_train, y_train)

## Predicting the Test set results
y_predict = regressor.predict(x_test)

## Steps of Backward Eliminations
# 1. Select a signficance level to stay in the model (e.g. SL = 0.05)
# 2. Fit the full model with all possible predictors
# 3. Consider the predictor with the highest P value. if P > SL, go to step 4, otherwise it's done
# 4. Remove the predictor
# 5. Fit model without his variable

## Building the optimal model using backward elimintaion
import statsmodels.api as sm
x = np.append(arr=np.ones((50, 1)).astype(int), values=x, axis=1)

# Step 3 - 5
x_opt = x
regressor_OLS = sm.OLS(endog=y, exog=x_opt).fit()
print(regressor_OLS.summary())

x_opt = x[:, [0, 1, 3, 4, 5]]
regressor_OLS = sm.OLS(endog=y, exog=x_opt).fit()
print(regressor_OLS.summary())

x_opt = x[:, [0, 1, 3, 4, 5]]
regressor_OLS = sm.OLS(endog=y, exog=x_opt).fit()
print(regressor_OLS.summary())

x_opt = x[:, [0, 3, 4, 5]]
regressor_OLS = sm.OLS(endog=y, exog=x_opt).fit()
print(regressor_OLS.summary())

x_opt = x[:, [0, 3, 5]]
regressor_OLS = sm.OLS(endog=y, exog=x_opt).fit()
print(regressor_OLS.summary())

x_opt = x[:, [0, 3]]
regressor_OLS = sm.OLS(endog=y, exog=x_opt).fit()
print(regressor_OLS.summary())

## Automatic Backward Elimination
def backwardElimination(x, sl):
    numVars = len(x[0])
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x).fit()
        maxVar = max(regressor_OLS.pvalues).astype(float)
        if maxVar > sl:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j].astype(float) == maxVar):
                    x = np.delete(x, j, 1)
    print(regressor_OLS.summary())
    return x
 
SL = 0.05
x_opt = x[:, [0, 1, 2, 3, 4, 5]]
x_modeled = backwardElimination(x_opt, SL)